xds/internal/balancer/clusterimpl/clusterimpl.go

```diff
@@ -0,0 +1,315 @@
+		if cib.xdsC != nil {
+			loadStore, cib.cancelLoadReport = cib.xdsC.ReportLoad(cib.lrsServerName)
+		}
+		cib.loadWrapper.UpdateLoadStore(loadStore)
+	}
+
+	return nil
+}
+
+func (cib *clusterImplBalancer) UpdateClientConnState(s balancer.ClientConnState) error {
+	newConfig, ok := s.BalancerConfig.(*lbConfig)
+	if !ok {
+		return fmt.Errorf("unexpected balancer config with type: %T", s.BalancerConfig)
+	}
+
+	// Update load reporting config. This needs to be done before updating the
+	// child policy because we need the loadStore from the updated client to be
+	// passed to the ccWrapper.
+	if err := cib.updateLoadStore(newConfig); err != nil {
+		return err
+	}
+
+	// Compare new drop config. And update picker if it's changed.
+	var updatePicker bool
+	if cib.config == nil || !reflect.DeepEqual(cib.config.DropCategories, newConfig.DropCategories) {
+		cib.drops = make([]*dropper, 0, len(newConfig.DropCategories))
+		for _, c := range newConfig.DropCategories {
+			cib.drops = append(cib.drops, newDropper(c))
+		}
+		updatePicker = true
+	}
+
+	// Compare cluster name. And update picker if it's changed, because circuit
+	// breaking's stream counter will be different.
+	if cib.config == nil || cib.config.Cluster != newConfig.Cluster {
+		cib.requestCounter = xdsclient.GetServiceRequestsCounter(newConfig.Cluster)
+		updatePicker = true
+	}
+	// Compare upper bound of stream count. And update picker if it's changed.
+	// This is also for circuit breaking.
+	var newRequestCountMax uint32 = 1024
+	if newConfig.MaxConcurrentRequests != nil {
+		newRequestCountMax = *newConfig.MaxConcurrentRequests
+	}
+	if cib.requestCountMax != newRequestCountMax {
+		cib.requestCountMax = newRequestCountMax
+		updatePicker = true
+	}
+
+	if updatePicker {
+		cib.pickerUpdateCh.Put(&dropConfigs{
+			drops:          cib.drops,
+			requestCounter: cib.requestCounter,
+		})
+	}
+
+	// If child policy is a different type, recreate the sub-balancer.
+	if cib.config == nil || cib.config.ChildPolicy.Name != newConfig.ChildPolicy.Name {
+		bb := balancer.Get(newConfig.ChildPolicy.Name)
+		if bb == nil {
+			return fmt.Errorf("balancer %q not registered", newConfig.ChildPolicy.Name)
+		}
+		if cib.childLB != nil {
+			cib.childLB.Close()
+		}
+		cib.childLB = bb.Build(cib, balancer.BuildOptions{})
+	}
+	cib.config = newConfig
+
+	// Addresses and sub-balancer config are sent to sub-balancer.
+	return cib.childLB.UpdateClientConnState(balancer.ClientConnState{
+		ResolverState:  s.ResolverState,
+		BalancerConfig: cib.config.ChildPolicy.Config,
+	})
+}
+
+func (cib *clusterImplBalancer) ResolverError(err error) {
+	if cib.childLB != nil {
+		cib.childLB.ResolverError(err)
+	}
+}
+
+func (cib *clusterImplBalancer) UpdateSubConnState(sc balancer.SubConn, s balancer.SubConnState) {
+	if cib.childLB != nil {
+		cib.childLB.UpdateSubConnState(sc, s)
+	}
+}
+
+func (cib *clusterImplBalancer) Close() {
+	if cib.childLB != nil {
+		cib.childLB.Close()
+		cib.childLB = nil
```

Comment:
Related to whether we need to check `closed.HasFired()` in methods. If we do that, we don't need to set this to `nil` here, right?